SCRIPTS/DSH/dsh-1.21/dsh/dsh.1.in
.\"  Copyright (c) 2004, 2005, Larry Lile <lile@users.sourceforge.net>
.\"  All rights reserved.
.\"
.\"  Redistribution and use in source and binary forms, with or without
.\"  modification, are permitted provided that the following conditions
.\"  are met:
.\"  1. Redistributions of source code must retain the above copyright
.\"     notice unmodified, this list of conditions, and the following
.\"     disclaimer.
.\"  2. Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\"
.\"  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\"  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\"  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\"  ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\"  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\"  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\"  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\"  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\"  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\"  SUCH DAMAGE.
.\"
.\"  $Id: dsh.1.in,v 1.4 2005/07/20 00:12:23 lile Exp $
.\"
.Dd July 6, 2005
.Dt DSH 1
.Os "dsh (@VERSION@)"
.Sh NAME
.Nm dsh
.Nd distributed shell
.Sh SYNOPSIS
.Nm
.Op Fl q
.Pp
.Nm
.Op Fl h
.Pp
.Nm
.Op Fl i
.Op Fl v
.Op Fl c
.Op Fl t Ar csh|sh
.Op Fl d Li | Fl D
.Op Fl l Ar login_name
.Op Fl w Ar {host_name|@netgroup},... | Fl
.Op Fl f Ar fanout_value
.Op Fl o Dq Ar flags_and_parms
.Op Fl Y Ar domain
.Op command
.Sh DESCRIPTION
.Nm Dsh
executes commands in parallel on the machines specified in the working
collective.
.Nm Dsh
works in two distinct modes, batch mode where the command to be
executed is specified on the command line or interactive mode where a
command line shell will be presented to the user.
.Pp
The working collective can be specified in two ways, either on the command
line using
.Fl w
argument where hosts and netgroups can be specified in a comma separated
list.  Alternately the working collective can be specified on stdin using
the 
.Fl w Fl
with each host or netgroup on a line by itself.  The second method for
specifying the working collective is using the
.Ev WCOLL
environment variable.  The
.Ev WCOLL
environment variable should be set to the name of a file containing the list
of host names and netgroups listed on separate lines.  Blank lines and lines
beginning with # will be ignored.
.Pp
The path used on remote systems for the command is determined by the
.Ev DSHPATH
environment variable.  If the
.Ev DSHPATH
environment variable is not set, the default path,
.Em /usr/ucb:/bin:/usr/bin:. ,
will be used.  If
.Ev DSHPATH is defined but empty then no
.Ev PATH
will be set on the remote system, this is useful for systems with
non standard shells.  Otherwise if
.Ev DSHPATH
is set then the
.Ev PATH
environment variable will be set on the remote system to the value in
.Ev DSHPATH
using the appropriate shell syntax.
.Pp
The number of machines that the command will be executed on in parallel is
determined by the fanout value.  If commands are to be run serially on all
of the hosts in the working collective a fanout value of 1 should be specified.
The fanout value can be specified by using either the 
.Ev FANOUT environment variable or the
.Fl f
command line option, if neither of these is specified the default fanout value
of 64 will be used.  Large values for fanout should be used carefully, most
remote shell utilities require reserved TCP/IP ports and these are a limited
resource.  If large fanouts are specified or multiple users are using
.Nm
then the reserved TCP/IP ports can be easily exhausted causing failures.
.Pp
The exit values for remote commands are displayed in messages from
.Nm
if they are nonzero, a nonzero exit value from the remote command indicates
that the command was not successful.
.Nm
will remove hosts from the working collective if the remote command is
unsuccessful unless the
.Fl c
command line option is specified.
.Pp
.Bd -ragged -offset indent
.Em Note :
Not all remote shells return the exit status of the command properly and some
failures may not be detected.
.Ed
.Pp
.Nm
does not work with interactive commands.  The user must ensure that the
interactive commands can be run without prompts before attempting to use
them with
.Nm .
.Pp
The standard output and standard error for
.Nm
is the standard output and standard error, respectively, of the remote
command.  Each line of output is prefixed with the host name of the host from
which that output was received.  A ":" is used to separate the host name from
the output.
.Pp
.Pp
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl q
Displays the hosts in the current working collective, derived from either the
.Fl w
option or the
.Ev WCOLL
environment variable.  Also the current fanout value is displayed,
determined from either the
.Fl f 
option or the
.Ev FANOUT
environment variable.
.It Fl h
Displays the usage information.
.It Fl i
When set
.Nm
will display the working collective and fanout values before each
command is issued.
.It Fl v
Hosts are verified using ping before being added to the working collective.
If a host is not responding it will not be added to the collective.
.It Fl c
If set
.Nm
will continue to send commands to hosts that are not responding or have
returned a non-zero exit code.  If this flag is not set
.Nm
will remove hosts that are not responding or have returned a non-zero
exit code for the remainder of the session.
.It Fl t Ar sh|csh
Specifies the target shell syntax.  This flag is used to create the syntax
of the internal environment variables passed with the command.  Supported
shells are
.Ic sh
and
.Ic csh .
The default is
.Ic csh .
.It Fl d
Causes
.Nm
to set the
.Fl f
flag on the rsh command.  This option is only relevant for
.Ic rsh .
.It Fl D
Causes
.Nm
to set the
.Fl F
flag on the rsh command.  This option is only relevant for
.Ic rsh .
.It Fl l Ar login_name
sets the remote user name to 
.Ar login_name .
Otherwise, the  remote user name will be the same as the local user name.
.It Fl w Ar {host_name|@netgroup},... | Fl
Specifies one or more host names or netgroups, separated by commas.  No
whitespace is allowed.  Netgroups are designated by prefixing the netgroup
name with an @ symbol.
.Pp
If
.Fl w Fl
is specified, the working collective will be read one line at a time
from stdin.  Host names or netgroups should be specified one per line
until finished.  When finished
.Sy <Ctrl-D>
.It Fl f Ar fanout_value
Specifies the number of hosts in the collective to execute the command
in parallel.  The default value is 64.
.It Fl o Ar flags_and_parms
Allows additional arguments to be specified to the remote command program.
The flags and parameters will need to be surrounded with single or double
quotes.  The 
.Fl o
parameter can be specified multiple time to build up the argument list
as needed.
.It Fl Y Ar domain
Specify the NIS domain for netgroup lookups if it is not the default
NIS domain.
.It Op Ar command
The command to be executed on the hosts in the working collective.
.El
.Pp
.Sh ENVIRONMENT
The behavior of
.Nm
can be altered using the following environment variables.
.Bl -tag -width Ds
.It Ev WCOLL
Specifies a working collective file.
.It Ev DSHPATH
Specifies a
.Ev PATH
to be set for the remote command environment.  If set, but empty no
.Ev PATH
will be set.
.It Ev DSH_REMOTE_CMD
The command to be used for remote execution, this is typically
.Xr rsh
or
.Xr ssh
but can be any suitable remote execution command.
.It Ev RCMD_PGM
Defines the type of remote execution command, either rsh or secrshell.
.It Ev HN_METHOD
Can be set to either
.Dq reliable
or
.Dq initial .
The default host name method is initial, which is the name that was specified
by the user.  Setting the host name method to reliable will result in the canonical
fully qualified host name being used.
.It Ev DSH_COLLATE
Defining
.Ev DSH_COLLATE will cause
.Nm
to hold output for each host until the remote command is finished before displaying
the output.  Use of 
.Ev DSH_COLLATE
is not recommended, dshbak should be used instead.
.It Ev DSH_DEBUG
Defining
.Ev DSH_DEBUG
will cause copious debugging information for
.Nm
to be displayed.
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa @bindir@/dshbak
A utility provided to reformat the output of
.Nm
by host or alternately to collate identical output from groups of hosts.
.It working collective file
A file containing host names or netgroup names, one per line, used to
define a working collective.
.El
.Sh EXAMPLES
.Sh DIAGNOSTICS
.Sh COMPATIBILITY
.Nm Dsh
is compatible with the IBM PSSP
.Nm
command, with the following exceptions.
.Fl a , Fl G ,
and
.Fl N
are not implemented as they are specific to the System Data Repository.
.Sh AUTHORS
.Nm Dsh
first appeard in IBM's Parallel System Support Package for AIX.
.Sh BUGS
.Bl -item -width Ds
.It
The names of the environment variables are inconsistent, but match the
original implementation.  The additional environment variables not defined
by IBM are named consistently.
.It
This man page is very incomplete.
.El
.Sh COPYRIGHT
Copyright \(co 2005 Larry Lile <lile@users.sourceforge.net>
.br
This is free software; see the source for copying conditions.  There
is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.
.Sh SEE ALSO
.Xr rsh 1 ,
.Xr ssh 1 ,
.Xr dshbak 1
.Pp


SCRIPTS/DSH/dsh-1.21/dsh/dsh.c
/*
 * Copyright (c) 2004, 2005, Larry Lile <lile@users.sourceforge.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice unmodified, this list of conditions, and the following
 *    disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: dsh.c,v 1.25 2005/10/18 01:20:45 lile Exp $
 */

#include "includes.h"

RCSID("$Id: dsh.c,v 1.25 2005/10/18 01:20:45 lile Exp $");

#if !defined(RSH_DFLT)
#define RSH_DFLT "/usr/bin/rsh"
#endif /* !defined RSH_DFLT */

#if !defined(SSH_DFLT)
#define SSH_DFLT "/usr/bin/ssh"
#endif /* !defined SSH_DFLT */

#if !defined(DSHPATH_DFLT)
#define DSHPATH_DFLT "/usr/ucb:/bin:/usr/bin:."
#endif /* !defined DSHPATH_DFLT */

#define MAX_FANOUT(x)	((x - 3) / 3)
#define TIMEOUT		180

unsigned int	 qflag = 0,
		 iflag = 0,
		 vflag = 0,
		 cflag = 0,
		 tflag = 1, /* XXX Default shell from configure */
		 dflag = 0,
		 Dflag = 0,
		 fflag = 64;

char		*lflag = (char *)NULL,
		*wflag = (char *)NULL,
		*oflag = (char *)NULL,
		*Yflag = (char *)NULL;

int		 signaled = 0,
		 sigcount = 0;

unsigned int	 collate = 0;

extern int 	 errno;

char		*__ypdomain = NULL;

char		**remote_argv = (char **)NULL,
		 *remote_host = (char *)NULL,
		**remote_cmdv = (char **)NULL;
int		  remote_argc = 0;

void
push_remote_argv(char *argument)
{
	char **tmp;

	/* +2 because we keep the remote_argv null terminated  */
	if ((tmp = (char **)malloc(sizeof(char *) * (remote_argc + 2))) == NULL)
	{
		perror("Unable to allocate remote arguments - malloc");
		exit(1);
	}
	memset(tmp, '\0', sizeof(char *) * (remote_argc + 2));

	if (remote_argv)
	{
		memcpy(tmp, remote_argv, remote_argc * sizeof(char *));
		free(remote_argv);
	}
	remote_argv = tmp;

	if (argument)
		remote_argv[remote_argc] = strdup(argument);

	remote_argc++;
}

void
sig_handler(int signum)
{
	debug_msg("sig_handler(%d)\n", signum);
	signaled = signum;
	sigcount++;
	if (sigcount > 5)
	{
		print_error("Something must be hung, exiting...\n");
		exit(1);
	}
}

char *
clean_string(char *string)
{
	char	*s = string,
		*t;

	while (s && isspace(*s))
		s++;

	if (*s == '#')
		return (char *)NULL;

	if (!s)
		return s;

	t = s + strlen(s) - 1;
	while (t > s && isspace(*t))
		t--;
	*++t = '\0';

	return s;
}

int
execute_command(int quiet, char *hn_method, int timeout, int argc, char **argv)
{
	int		 rc = 0;
	struct member	*ptr;
	int		 watchdog = timeout,
			 watchdogs_set = 0,
			 queue_watchdog = 0;

	signaled = 0;
	sigcount = 0;

	if (iflag)
	{
		char **cmdp = remote_cmdv;
		collective_display();
		printf("command: ");
		while(*cmdp)
			printf("%s ", *cmdp++);
		printf("\n");
	}

	queue_max_slots(fflag);
	ptr = collective_get_next(NULL);
	while (((ptr && !signaled) || queue_load()))
	{
		time_t now = time(NULL);
		if (queue_slots() == 0)
		{
			if (watchdog > now)
			{
				debug_msg("queue watchdog expired\n");
				queue_set_watchdogs(watchdog);
				watchdogs_set = 1;
			}
			if (!watchdog)
			{
				debug_msg("setting watchdog on queue\n");
				queue_watchdog = now + watchdog;
			}
		}
		if (queue_slots() && queue_watchdog)
		{
			debug_msg("clearing watchdog on queue\n");
			queue_watchdog = 0;
		}

		while (queue_slots() > 0 && (!signaled && ptr))
		{
			struct member	*next;
			next = collective_get_next(ptr);
			if (cflag || ptr->responding)
			{
				char *name = ptr->name;
				if (!strcmp(hn_method, "reliable"))
					name = ptr->reliable;
				queue_insert(ptr, name, argc, argv);
			} else {
				debug_msg("%s not responding, dropped from collective\n",
						ptr->name);
				collective_drop_member(ptr);
			}
			ptr = next;
		}
		rc += queue_run(quiet, signaled);

		if (!ptr && queue_load() && !watchdogs_set)
		{
			queue_set_watchdogs(watchdog);
			watchdogs_set = 1;
		}
	}

	if (signaled > 0 && queue_load() > 0)
	{
		debug_msg("queued processes to be killed: %d\n",
		    queue_load());
	}

	return rc;
}

int
main(int argc, char **argv)
{
	int		 ch,
			 i,
			 dshtimeout = TIMEOUT;
	char		*wcoll = (char *)NULL,
			*fanout,
			*dsh_remote_cmd = RSH_DFLT,
			*rcmd_pgm = "rsh",
			*hn_method = (char *)NULL,
			*dshpath = (char *)NULL,
			*prompt,
			*envp;

	char		*ping[] = { "ping", "-w", "4", NULL, NULL };

#if defined(RLIMIT_NOFILE) || defined(RLIMIT_NPROC)
	struct rlimit	 rl;
#endif /* defined(RLIMIT_NOFILE) || defined(RLIMIT_NPROC) */

	int		 rc = 0;

	if (geteuid() == 0)
		prompt = "dsh# ";
	else
		prompt = "dsh> ";

	if (getenv("DSH_DEBUG") != NULL)
		set_debug(1);
	debug_msg("DSH_DEBUG\n");

	rcmd_pgm = (envp = getenv("RCMD_PGM")) ? envp : "rsh";
	if (!strcmp(rcmd_pgm, "rsh"))
		dsh_remote_cmd = (envp = getenv("DSH_REMOTE_CMD")) ? envp : RSH_DFLT;
	else if (!strcmp(rcmd_pgm, "secrshell"))
		dsh_remote_cmd = (envp = getenv("DSH_REMOTE_CMD")) ? envp : SSH_DFLT;
	else
	{
		print_error("unknown RCMD_PGM %s\n", rcmd_pgm);
		exit(1);
	}
	debug_msg("DSH_REMOTE_CMD = %s\n", dsh_remote_cmd);
	debug_msg("RCMD_PGM = %s\n", rcmd_pgm);

	push_remote_argv(dsh_remote_cmd);

	hn_method = (envp = getenv("HN_METHOD")) ? envp : "initial_hostname";
	if (strcmp(hn_method, "initial_hostname") &&
	    strcmp(hn_method, "reliable"))
	{
		print_error("unknown HN_METHOD %s\n", hn_method);
		exit(1);
	}
	debug_msg("HN_METHOD = %s\n", hn_method);

	dshpath = (envp = getenv("DSHPATH")) ? envp : DSHPATH_DFLT;
	debug_msg("DSHPATH = %s\n", dshpath);

	if (getenv("DSH_TIMEOUT"))
		dshtimeout = atoi(getenv("DSH_TIMEOUT"));
	debug_msg("DSH_TIMEOUT = %d\n", dshtimeout);

	if (getenv("DSH_COLLATE") != NULL)
	{
		collate = 1;
		debug_msg("DSH_COLLATE\n");
	}

#if defined(RLIMIT_NOFILE)
	if (getrlimit(RLIMIT_NOFILE, &rl) == -1)
	{
		perror("getrlimit");
	} else {
		debug_msg("RLIMIT_NOFILE = %d (%d)\n",
		    rl.rlim_cur, rl.rlim_max);
		debug_msg("maximum fanout: %d\n", MAX_FANOUT(rl.rlim_cur));
		/* XXX Reduce fflag if necessary */
	}
#endif /* defined(RLIMIT_NOFILE) */

#if defined(RLIMIT_NPROC)
	if (getrlimit(RLIMIT_NPROC, &rl) == -1)
	{
		perror("getrlimit");
	} else {
		debug_msg("RLIMIT_NPROC = %d (%d)\n",
		    rl.rlim_cur, rl.rlim_max);
		/* XXX Reduce fflag if necessary */
	}
#endif /* defined(RLIMIT_NPROC) */

	if ((fanout = getenv("FANOUT")) != NULL && atoi(fanout) > 0)
		fflag = atoi(fanout);

	while ((ch = getopt(argc, argv, "qhivct:dDl:w:f:o:")) != -1)
	{
		switch(ch)
		{
			case 'q':
				qflag = 1;
				break;
			case 'h':
				exit(usage(argv[0]));
				break;
			case 'i':
				iflag = 1;
				break;
			case 'v':
				vflag = 1;
				break;
			case 'c':
				cflag = 1;
				break;
			case 't':
				if (!strcmp(optarg, "sh") || !strcmp(optarg, "ksh"))
					tflag = 0;
				else if (!strcmp(optarg, "csh") || !strcmp(optarg, "tcsh"))
					tflag = 1;
				else
				{
					print_error("%s unknown shell syntax.\n", optarg);
					exit(1);
				}
				break;
			case 'd':
				dflag = 1;
				if (Dflag)
				{
					print_error("-D and -d cannot be used at the same time.\n");
					exit(1);
				}
				break;
			case 'D':
				Dflag = 1;
				if (dflag)
				{
					print_error("-D and -d cannot be used at the same time.\n");
					exit(1);
				}
				break;
			case 'l':
				lflag = strdup(optarg);
				break;
			case 'w':
				wflag = strdup(optarg);
				break;
			case 'f':
				fflag = atoi(optarg);
				break;
			case 'o':
				oflag = strdup(optarg);
				push_remote_argv(oflag);
				break;
			case 'Y':
				Yflag = strdup(optarg);
				__ypdomain = Yflag;
				break;
			default:
				exit(!usage(argv[0]));
				break;
		}
	}

	argc -= optind;
	argv += optind;

	if (lflag)
	{
		push_remote_argv("-l");
		push_remote_argv(lflag);
	}
	if (dflag)
		push_remote_argv("-f");
	if (Dflag)
		push_remote_argv("-F");

 	/* XXX First NULL argument becomes hostname */
	remote_host = remote_argv[remote_argc];
	push_remote_argv(NULL);

	if (strlen(dshpath))
	{
		char *tmp;
		if (tflag)
		{
			tmp = (char *)malloc(sizeof(char) *
				(strlen(dshpath) + strlen("setenv PATH ;") + 1) );
			sprintf(tmp, "setenv PATH %s;", dshpath);
		} else {
			tmp = (char *)malloc(sizeof(char) *
				(strlen(dshpath) + strlen("PATH=; export PATH;") + 1) );
			sprintf(tmp, "PATH=%s; export PATH;", dshpath);
		}
		push_remote_argv(tmp);
		free(tmp);
	}

	if (argc)
		for (i = 0; i < argc; i++)
			push_remote_argv(argv[i]);
	remote_cmdv = &remote_argv[remote_argc - argc];

	debug_msg("remote_argv = ");
	for (i = 0; i < remote_argc; i++)
		debug_msg("%s ", (remote_argv[i] ? remote_argv[i] : "(NULL)"));

	debug_msg("\nremote_argc = %d\n", remote_argc);
	
	if (!wflag && ((wcoll = getenv("WCOLL")) == NULL))
	{
		print_error("working collective not set.\n");
		exit(1);
	}

	if (wflag)
	{
		if (*wflag != '-')
		{
			debug_msg("loading collective from -w\n");
			collective_parse(wflag, ',');
		} else {
			debug_msg("loading collective from stdin\n");
			collective_add_stdin(stdin);
		}
	} else {
		FILE	*file = stdin;
		debug_msg("loading collective from WCOLL=%s\n", wcoll);
		if ((file = fopen(wcoll, "r")) == NULL)
		{
			perror("fopen");
			return 1;
		}
		collective_add_file(file);
	}

	if (qflag)
		return collective_display();

	signal(SIGINT,  &sig_handler);
	signal(SIGQUIT, &sig_handler);
	signal(SIGABRT, &sig_handler);

	if (vflag)
	{
		execute_command(1, hn_method, 10, 4, ping);
	}

	if (argc == 0)
	{
#if defined(HAVE_READLINE)
		char *line;
		while (1)
		{
			char *s;

			signal(SIGINT,  SIG_DFL);
			signal(SIGQUIT, SIG_DFL);
			signal(SIGABRT, SIG_DFL);
			
			if ((line = readline(prompt)) == NULL)
				break;

			if ((s = clean_string(line)) == NULL)
				continue;

			signal(SIGINT,  &sig_handler);
			signal(SIGQUIT, &sig_handler);
			signal(SIGABRT, &sig_handler);
			
			if (s && strlen(s))
			{
				if (!strcmp(s, "exit"))
					break;
#if defined(HAVE_ADD_HISTORY)
				add_history(s);
#endif /* HAVE_ADD_HISTORY */
				if (*s == '!')
				{
					printf("local command: %s\n", s);
					continue;
				} else {
					remote_argv[remote_argc - 1] = strdup(s);
					rc = execute_command(0, hn_method, dshtimeout,
							remote_argc, remote_argv);
				}
			}
		}
		if (line == (char *)NULL)
			printf("\n");
#else
		printf("readline not availble\n");
#endif /* HAVE_READLINE */
	} else {
		rc = execute_command(0, hn_method, dshtimeout, remote_argc, remote_argv);
	}

	signal(SIGINT,  SIG_DFL);
	signal(SIGQUIT, SIG_DFL);
	signal(SIGABRT, SIG_DFL);

	for (i = 0; i < remote_argc; i++)
		if (remote_argv[i])
		{
			free(remote_argv[i]);
			remote_argv[i] = (char *)NULL;
		}
	free(remote_argv);
	remote_argv = (char **)NULL;

	free_collective();

	if (lflag)
	{
		free(lflag);
		lflag = (char *)NULL;
	}
	if (wflag)
	{
		free(wflag);
		wflag = (char *)NULL;
	}
	if (oflag)
	{
		free(oflag);
		oflag = (char *)NULL;
	}
	if (Yflag)
	{
		free(Yflag);
		Yflag = (char *)NULL;
	}

	return rc;
}
